include/benchmark/benchmark_api.h

```diff
@@ -227,6 +231,126 @@ inline BENCHMARK_ALWAYS_INLINE void DoNotOptimize(Tp const& value) {
+
+  Counter();
+  Counter(const char *name);
+  Counter(const char *name, double v);
+  Counter(const char *name, double v, uint32_t flags);
+  ~Counter();
+  Counter(Counter const& that);
+  Counter& operator= (Counter const& that);
+#ifdef BENCHMARK_HAS_CXX11
+  Counter(Counter && that);
+  Counter& operator= (Counter && that);
+#endif
+
+  void Set(double v) { value_ = v; }
+
+  void Finish(double cpu_time, double num_threads);
+
+  const char* Name() const { return name_; }
+  double Value() const { return value_; }
+
+  bool SameName(const char *n) const;
+
+  Counter& operator+= (double v) { value_ += v; return *this; }
+  Counter& operator-= (double v) { value_ -= v; return *this; }
+  Counter& operator*= (double v) { value_ *= v; return *this; }
+  Counter& operator/= (double v) { value_ /= v; return *this; }
+
+private:
+
+  // using raw C strings here to avoid inclusion of std::string in the public api
+  char name_buf_[16]; // strings smaller than 15 characters are stored here
+  char* name_mem_;    // whereas longer strings are stored here
+  size_t mem_size_;   // current size of name_mem_
+  char* name_;        // this points to the right one
+
+  double   value_;
+  uint32_t flags_;
+
+  void _SetName(const char *name);
+};
+
+
+
+class BenchmarkCounters {
+public:
+
+  typedef Counter        value_type;
+  typedef Counter const* const_iterator;
+
+public:
+
+  BenchmarkCounters();
+  ~BenchmarkCounters();
+  BenchmarkCounters(BenchmarkCounters const& that);
+  BenchmarkCounters& operator= (BenchmarkCounters const& that);
+#ifdef BENCHMARK_HAS_CXX11
+  BenchmarkCounters(BenchmarkCounters && that);
+  BenchmarkCounters& operator= (BenchmarkCounters && that);
+#endif
+
+  // Add/Set a counter.
+  size_t  Set(Counter const& c);
+  // Add/Set a counter.
+  size_t  Set(const char* n, double v) { return Set(n, v, Counter::kDefaults); }
+  // Add/Set a counter. When the counter already exists, f is ignored.
+  size_t  Set(const char* n, double v, uint32_t f);
+
+#ifdef BENCHMARK_HAS_CXX11
+  void    Set(std::initializer_list< Counter > il) { for(auto &c : il) { Set(c); } }
+#endif
+
+  // Returns the index where name is located, or size() if the name was not found.
+  size_t  Find(const char* name) const;
+
+  bool Exists(const char* name) const { return Find(name) < num_counters_; }
+
+  Counter & Get(size_t id);
+  Counter & Get(const char* name);
+  Counter const& Get(size_t id) const;
+  Counter const& Get(const char* name) const;
+
+  bool SameNames(BenchmarkCounters const& that) const;
+
+  void Sum(BenchmarkCounters const& that);
+  void Finish(double time, double num_threads);
+  void Clear();
+
+  size_t size() const { return num_counters_;  }
+
+  const_iterator begin() const { return counters_; }
+  const_iterator end () const { return counters_ + num_counters_; }
+
+private:
+
+  // using raw pointers here to avoid inclusion of std::vector in the public api
+  Counter * counters_;
+  size_t num_counters_;
+  size_t capacity_;
```

Comment:
NOTE: this fixes the <vector> include. I ended up using raw pointers instead of PIMPL. This was validated in valgrind and passed all tests so far.